src/hammer-vlsi/hammer_vlsi/submit_command.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# submit_command.py
+#
+
+from abc import ABCMeta, abstractmethod
+import atexit
+import subprocess
+import hammer_utils
+from typing import Callable, Iterable, List, NamedTuple, Optional, Dict, Any, Union
+from hammer_logging import HammerVLSIFileLogger, HammerVLSILogging, HammerVLSILoggingContext
+from hammer_config import HammerDatabase
+from functools import reduce
+
+class HammerSubmitCommand:
+
+    @abstractmethod
+    def submit(self, args: List[str], env: Dict[str,str], logger: HammerVLSILoggingContext, cwd: str = None) -> str:
+        """
+        Submit the job to the job submission system. This function MUST block until the command is complete.
+
+        :param args: Command-line to run; each item in the list is one token. The first token should be the command to run.
+        :param env: The environment variables to set for the command
+        :param logger: The logging context
+        :param cwd: Working directory (leave as None to use the current working directory).
+        :return: The command output
+        """
+        pass
+
+    @abstractmethod
+    def read_settings(self, settings: Dict[str, Any], tool_namespace: str) -> None:
+        """
+        Read the settings object (a Dict[str, Any]) into meaningful class variables
+
+        :param settings: A Dict[str, Any] comprising the settings for this command
+        :param tool_namespace: The namespace for the tool (useful for logging)
+        """
+        pass
+
+
+    @staticmethod
+    def get(tool_namespace: str, database: HammerDatabase) -> "HammerSubmitCommand":
```

Comment:
I almost feel like this method might as well live in HammerDriver...

Reply:
I personally like static factory methods living in their classes, but I don't have a strong opinion.